### PR TITLE
Add XSC analytics metrics capabilities

### PR DESCRIPTION
**Pull request by gailazar300**
- [x] The pull request is targeting the `dev` branch.
- [x] The code has been validated to compile successfully by running `go vet ./...`.
- [x] The code has been formatted properly using `go fmt ./...`.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-cli-security/actions/workflows/analysis.yml) passed.
- [x] All [tests](https://github.com/jfrog/jfrog-cli-security/actions/workflows/test.yml) have passed. If this feature is not already covered by the tests, new tests have been added.
- [x] All changes are detailed at the description. if not already covered at [JFrog Documentation](https://github.com/jfrog/documentation), new documentation have been added.

-----

Add analytics metrics capabilities for XSC.
* Adding new XSC server with analytics general events APIs.
* Adding new XSC tests section (for now XSC integration tests were tested and passed with d2c environment which has XSC, skipped in Github-action for now)
* All CLI audit commands will send a request for scanning to XSC instead of Xray from now on.

depends on:
https://github.com/jfrog/jfrog-client-go/pull/928
https://github.com/jfrog/jfrog-cli-core/pull/1165

### CODE REVIEW

File: utils/xscmanager.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package utils
+
+import (
+	"github.com/jfrog/jfrog-cli-core/v2/utils/config"
+	clientconfig "github.com/jfrog/jfrog-client-go/config"
+	"github.com/jfrog/jfrog-client-go/xsc"
+)
+
+func CreateXscServiceManager(serviceDetails *config.ServerDetails) (*xsc.XscServicesManager, error) {
```
Comment (eranturgeman) [inline utils/xscmanager.go:9]: Please change argument name to serverDetails instead of serviceDetails because server and service mean 2 different things in this context and its best to differentiate between them

File: commands/audit/jas/common.go [hunk 2]
```diff
@@ -274,3 +274,30 @@ func GetExcludePatterns(module jfrogappsconfig.Module, scanner *jfrogappsconfig.
...
+func ResetAnalyticsMetricsDataForAnalyzerManager() {
+	err := os.Setenv("AM_PACKAGE_MANAGER", "")
+	if err != nil {
+		log.Debug(fmt.Sprintf("failed setting AM_PACKAGE_MANAGER as environment variable. Cause: %s", err.Error()))
+	}
+	err = os.Setenv("AM_LANGUAGE", "")
+	if err != nil {
+		log.Debug(fmt.Sprintf("failed setting AM_LANGUAGE as environment variable. Cause: %s", err.Error()))
+	}
+}
```
Comment (attiasas) [inline commands/audit/jas/common.go:303]:
```suggestion
func SetAnalyticsMetricsDataForAnalyzerManager(technologies []coreutils.Technology) func() {
	if len(technologies) != 1 {
        // Only report analytics for one technology at a time.
		return
	}
	technology := technologies[0]
	resetAnalyzerManagerPackageManagerVar := SetEnvWithResetCallback(jfPacakgeManagerEnvVariable, technology.String())
	resetAnalyzerManagerLanguageVar := SetEnvWithResetCallback(jfLanguageEnvVariable, technology.Language)
	return func() {
		resetAnalyzerManagerPackageManagerVar()
		resetAnalyzerManagerLanguageVar()
	}
}
```

* create const variables at `analyzermanager.go` and use it
* You can return a func when setting the variables that will reset them
* add general `SetEnvWithResetCallback` to `utils.go` and use it:
```
func SetEnvWithResetCallback(key, value string) func() {
	oldValue, exist := os.LookupEnv(key)
    errMsg := "failed %s %s as environment variable. Cause: %s"

    if err := os.Setenv(key, value); err != nil {
        log.Debug(fmt.Sprintf(errMsg, "setting", key, err.Error()))
        return
    }

	if exist {
		return func() {
            if err := os.Setenv(key, oldValue); err != nil {
                log.Debug(fmt.Sprintf(errMsg, "setting", key, err.Error()))
            }
		}
	}

	return func() {
		if err := os.Unsetenv(key); err != nil {
            log.Debug(fmt.Sprintf(errMsg, "unsetting", key, err.Error()))
        }
	}
}
```

File: commands/audit/audit.go [hunk 3]
```diff
@@ -98,10 +103,13 @@ func (auditCmd *AuditCommand) Run() (err error) {
 		SetGraphBasicParams(auditCmd.AuditBasicParams).
 		SetThirdPartyApplicabilityScan(auditCmd.thirdPartyApplicabilityScan)
 	auditParams.SetIsRecursiveScan(isRecursiveScan).SetExclusions(auditCmd.Exclusions())
+
+	auditCmd.analyticsMetricsService.AddGeneralEventAndSetMsi(auditParams.xrayGraphScanParams)
```
Comment (attiasas) [inline commands/audit/audit.go:107]: this is a blocking method, consider running this optional operations in diff rotine

File: commands/audit/audit.go [hunk 4]
```diff
@@ -171,10 +179,7 @@ func RunAudit(auditParams *AuditParams) (results *xrayutils.Results, err error)
 		errGroup.Go(utils.DownloadAnalyzerManagerIfNeeded)
 	}
 
-	if auditParams.xrayGraphScanParams.XscGitInfoContext != nil {
-		if err = xrayutils.SendXscGitInfoRequestIfEnabled(auditParams.xrayGraphScanParams, xrayManager); err != nil {
-			return nil, err
-		}
+	if auditParams.xrayGraphScanParams.MultiScanId != "" {
 		results.MultiScanId = auditParams.xrayGraphScanParams.MultiScanId
 	}
```
Comment (attiasas) [inline commands/audit/audit.go:184]:
```suggestion
	results.MultiScanId = auditParams.MultiScanId
```
MultiScanId should be stored at auditParams, at this point of adding MSI to every audit scan we can just set the value from the params to the one at the result

File: utils/analyticsmetrics.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		return
+	}
+	log.Debug(fmt.Sprintf("New General event added successfully. multi_scan_id %s", ams.GetMsi()))
+
+	if err = os.Setenv(jfMsiEnvVariable, ams.GetMsi()); err != nil {
+		// Not a fatal error, if not set the scan will not be shown at the XSC UI, should not fail the scan.
+		log.Debug(fmt.Sprintf("failed setting MSI as environment variable. Cause: %s", err.Error()))
+	}
+	// Before running the audit command, set the msi so the sca scan will be performed on the xsc rather than on the xray server.
+	params.MultiScanId = ams.GetMsi()
```
Comment (attiasas) [inline utils/analyticsmetrics.go:97]: This method should return the msi and it should be set at the AuditBasicParams (that will set the value of XrayGraphScanParams)
This env var is used to pass the value to the analyzer manager only

File: utils/analyticsmetrics_test.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	originalReportUsage := os.Getenv(coreutils.ReportUsage)
+	defer func() {
+		err := os.Setenv(coreutils.ReportUsage, originalReportUsage)
+		assert.NoError(t, err)
+	}()
+
+	err := os.Setenv(coreutils.ReportUsage, "")
+	assert.NoError(t, err)
+	err = os.Setenv(jfMsiEnvVariable, "")
+	assert.NoError(t, err)
```
Comment (attiasas) [inline utils/analyticsmetrics_test.go:42]:
```suggestion
	cleanReportEnv := clientTests.SetEnvWithCallbackAndAssert(t, coreutils.ReportUsage, "")
    cleanMSIEnv := clientTests.SetEnvWithCallbackAndAssert(t, jfMsiEnvVariable, "")
```

Use clientTests.SetEnvWithCallbackAndAssert

File: utils/analyticsmetrics_test.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	err = os.Setenv(jfMsiEnvVariable, "msi")
+	assert.NoError(t, err)
+	assert.False(t, am.calcShouldReportEvents())
+
+	// JFROG_CLI_REPORT_USAGE is false.
+	err = os.Setenv(jfMsiEnvVariable, "")
+	assert.NoError(t, err)
+	err = os.Setenv(coreutils.ReportUsage, "false")
+	assert.NoError(t, err)
+	assert.False(t, am.calcShouldReportEvents())
```
Comment (attiasas) [inline utils/analyticsmetrics_test.go:72]: Define and use Test cases:

testCases := []struct {
    TestName string
    Version string
    expectedResult bool
}{
    ...
}

File: utils/analyticsmetrics_test.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	}()
+
+	// Successful flow.
+	mockServer, serverDetails := xscServer(t, AnalyticsMetricsMinXscVersion)
+	defer mockServer.Close()
+	am := NewAnalyticsMetricsService(serverDetails)
+	params := services.XrayGraphScanParams{}
+	am.AddGeneralEventAndSetMsi(&params)
+	assert.NotEmpty(t, am.GetMsi())
+	assert.Equal(t, params.MultiScanId, am.GetMsi())
```
Comment (attiasas) [inline utils/analyticsmetrics_test.go:88]:
```suggestion
	assert.NotEmpty(t, am.GetMsi())
	assert.Equal(t, params.MultiScanId, am.GetMsi())
```

File: utils/analyticsmetrics_test.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	assert.Equal(t, am.GetMsi(), os.Getenv(jfMsiEnvVariable))
+
+	// In case cli should not report analytics, verify that request won't be sent.
+	am.shouldReportEvents = false
+	am.SetMsi("test-msi")
+	am.AddGeneralEventAndSetMsi(&params)
+	assert.Equal(t, "test-msi", am.GetMsi())
+}
+
+func TestCreateAuditResultsFromXscAnalyticsBasicGeneralEvent(t *testing.T) {
```
Comment (attiasas) [inline utils/analyticsmetrics_test.go:98]: can be implemented in test cases, similar code is called.

File: utils/analyticsmetrics_test.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+		}
+		if r.RequestURI == "/xsc/api/v1/event" {
+			if r.Method == http.MethodPost {
+				w.WriteHeader(http.StatusCreated)
+				_, err := w.Write([]byte(fmt.Sprintf(`{"multi_scan_id": "%s"}`, testMsi)))
+				if err != nil {
+					return
+				}
+			}
+		}
```
Comment (attiasas) [inline utils/analyticsmetrics_test.go:157]: what about different in Methods `PUT` and `POST`?

File: go.mod [hunk 11]
```diff
@@ -98,8 +98,8 @@ require (
 	gopkg.in/warnings.v0 v0.1.2 // indirect
 )
 
-// replace github.com/jfrog/jfrog-cli-core/v2 => github.com/jfrog/jfrog-cli-core/v2 dev
+replace github.com/jfrog/jfrog-cli-core/v2 => github.com/gailazar300/jfrog-cli-core/v2 v2.0.0-20240331095546-a03b2d7f67da
 
-// replace github.com/jfrog/jfrog-client-go => github.com/jfrog/jfrog-client-go dev
+replace github.com/jfrog/jfrog-client-go => github.com/gailazar300/jfrog-client-go v0.18.1-0.20240331091256-e2e7cd419323
```
Comment (attiasas) [inline go.mod:104]: make sure to point to valid remote branch
don't forget to update before merge to point to the dev branch

**Review by attiasas — CHANGES_REQUESTED** [review]
Add MultiScanId to `AuditBasicParams` and use it in as input if MSI was supplied (at Frogbot you can set this param)

Use format tools for the logs.

Describe at the PR description the changes.

More tests are needed

File: go.mod (hunk 11 above)
Comment (gailazar300) [reply go.mod:104]: I have to change the go.mod to run tests with my core and client changes in github-action.

File: utils/analyticsmetrics_test.go (hunk 8 above)
Comment (gailazar300) [reply utils/analyticsmetrics_test.go:88]: what did you mean?

File: utils/analyticsmetrics_test.go (hunk 7 above)
Comment (gailazar300) [reply utils/analyticsmetrics_test.go:72]: this structure is not suitable for this specific test.

File: commands/audit/jas/common.go [hunk 12]
```diff
@@ -24,7 +25,9 @@ import (
 )
 
 const (
-	NodeModulesPattern = "**/*node_modules*/**"
+	NodeModulesPattern          = "**/*node_modules*/**"
+	jfPackageManagerEnvVariable = "AM_PACKAGE_MANAGER"
+	jfLanguageEnvVariable       = "AM_LANGUAGE"
```
Comment (attiasas) [inline commands/audit/jas/common.go:30]: maybe add it to analyzermanager.go instead? the other env var for it defifned there

File: commands/audit/jas/common.go [hunk 13]
```diff
@@ -274,3 +277,30 @@ func GetExcludePatterns(module jfrogappsconfig.Module, scanner *jfrogappsconfig.
...
+		resetAnalyzerManagerLanguageVar()
+		resetAnalyzerManageJfMsiVar()
+	}
+}
+
+func ResetAnalyticsMetricsDataForAnalyzerManager() {
+	err := os.Setenv("AM_PACKAGE_MANAGER", "")
+	if err != nil {
+		log.Debug(fmt.Sprintf("failed setting AM_PACKAGE_MANAGER as environment variable. Cause: %s", err.Error()))
+	}
```
Comment (attiasas) [inline commands/audit/jas/common.go:301]:
```suggestion
	err := os.Setenv(jfPackageManagerEnvVariable, "")
	if err != nil {
		log.Debug(fmt.Sprintf("failed setting %s as environment variable. Cause: %s", jfPackageManagerEnvVariable, err.Error()))
	}
```

File: commands/audit/jas/common.go [hunk 14]
```diff
@@ -274,3 +277,30 @@ func GetExcludePatterns(module jfrogappsconfig.Module, scanner *jfrogappsconfig.
...
+	err = os.Setenv("AM_LANGUAGE", "")
+	if err != nil {
+		log.Debug(fmt.Sprintf("failed setting AM_LANGUAGE as environment variable. Cause: %s", err.Error()))
+	}
```
Comment (attiasas) [inline commands/audit/jas/common.go:305]:
```suggestion
	err = os.Setenv(jfLanguageEnvVariable, "")
	if err != nil {
		log.Debug(fmt.Sprintf("failed setting %s as environment variable. Cause: %s", jfLanguageEnvVariable, err.Error()))
	}
```

File: commands/audit/jasrunner.go [hunk 15]
```diff
@@ -30,6 +30,8 @@ func runJasScannersAndSetResults(scanResults *utils.Results, directDependencies
 	if progress != nil {
 		progress.SetHeadlineMsg("Running applicability scanning")
 	}
+	// Set environments variables for analytics in analyzers manager.
+	defer jas.SetAnalyticsMetricsDataForAnalyzerManager(msi, scanResults.GetScaScannedTechnologies())
```
Comment (attiasas) [inline commands/audit/jasrunner.go:34]: defer will only set the value after `runJasScannersAndSetResults` is done, after running the Applicability scan...

File: tests/utils/test_utils.go [hunk 16]
```diff
@@ -16,11 +16,12 @@ import (
...
 
-func InitSecurityTest(t *testing.T, minVersion string) {
+func InitSecurityTest(t *testing.T, xrayMinVersion, xscMinVersion string) {
 	if !*configTests.TestSecurity {
 		t.Skip("Skipping Security test. To run Security test add the '-test.security=true' option.")
 	}
-	ValidateXrayVersion(t, minVersion)
+	ValidateXrayVersion(t, xrayMinVersion)
+	ValidateXscVersion(t, xscMinVersion)
 }
```
Comment (attiasas) [inline tests/utils/test_utils.go:25]: check for ValidateXscVersion can be added at the specific test that uses it.
no need to add the check for all tests...

Comment (gailazar300) [reply tests/utils/test_utils.go:25]: there is no separation between xray and XSC tests, there are all security tests, and as far as we get more APIs will approach XSC instead of xray.

File: xsc_test.go [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+	// Event creation and addition information.
+	assert.Equal(t, "cli", event.Product)
+	assert.Equal(t, 1, event.EventType)
+	assert.NotEmpty(t, event.EventStatus)
+	assert.NotEmpty(t, event.AnalyzerManagerVersion)
+	// The information that was added after updating the event with the scan's results.
+	assert.NotEmpty(t, event.TotalScanDuration)
+	assert.True(t, event.TotalFindings > 0)
+}
```
[the remark this answers is not in the document]
Comment (gailazar300) [reply xsc_test.go:46]: There should be a separation between the regular audit tests and XSC tests, we are testing specific XSC APIs so there should be a separate test section for that and for more XSC tests to come.
This test is specifically restricted by AnalyticsMetricsMinXscVersion, all other audit tests should run without considering XSC.

File: commands/audit/jas/common.go [hunk 18]
```diff
@@ -274,3 +275,39 @@ func GetExcludePatterns(module jfrogappsconfig.Module, scanner *jfrogappsconfig.
...
+	resetAnalyzerManageJfMsiVar, err := clientutils.SetEnvWithResetCallback(utils.JfMsiEnvVariable, msi)
+	if err != nil {
+		log.Debug(fmt.Sprintf(errMsg, "setting", utils.JfMsiEnvVariable, err.Error()))
+	}
+	if len(technologies) != 1 {
+		// Only report analytics for one technology at a time.
+		return func() {
+			if resetAnalyzerManageJfMsiVar() != nil {
+				log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfMsiEnvVariable, err.Error()))
+			}
```
Comment (attiasas) [inline commands/audit/jas/common.go:290]:
```suggestion
			if err = resetAnalyzerManageJfMsiVar(); err != nil {
				log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfMsiEnvVariable, err.Error()))
			}
```
don't ignore err

File: commands/audit/jas/common.go [hunk 19]
```diff
@@ -274,3 +275,39 @@ func GetExcludePatterns(module jfrogappsconfig.Module, scanner *jfrogappsconfig.
...
+	return func() {
+		if resetAnalyzerManageJfMsiVar() != nil {
+			log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfMsiEnvVariable, err.Error()))
+		}
+		if resetAnalyzerManagerPackageManagerVar() != nil {
+			log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfPackageManagerEnvVariable, err.Error()))
+		}
+		if resetAnalyzerManagerLanguageVar() != nil {
+			log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfLanguageEnvVariable, err.Error()))
+		}
```
Comment (attiasas) [inline commands/audit/jas/common.go:311]:
```suggestion
		if err = resetAnalyzerManageJfMsiVar(); err != nil {
			log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfMsiEnvVariable, err.Error()))
		}
		if err = resetAnalyzerManagerPackageManagerVar(); err != nil {
			log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfPackageManagerEnvVariable, err.Error()))
		}
		if err = resetAnalyzerManagerLanguageVar(); err != nil {
			log.Debug(fmt.Sprintf(errMsg, "restoring", utils.JfLanguageEnvVariable, err.Error()))
		}
```
don't ignore err

File: commands/audit/jas/common_test.go [hunk 20]
```diff
@@ -88,3 +90,37 @@ func TestAddScoreToRunRules(t *testing.T) {
...
+		{name: "One valid technology", args: args{msi: "msi", technologies: []coreutils.Technology{coreutils.Maven}}, want: func() {
+			assert.Equal(t, string(coreutils.Maven), os.Getenv(utils.JfPackageManagerEnvVariable))
+			assert.Equal(t, string(utils.Java), os.Getenv(utils.JfLanguageEnvVariable))
+			assert.Equal(t, "msi", os.Getenv(utils.JfMsiEnvVariable))
+		}},
+		{name: "Multiple technologies", args: args{msi: "msi", technologies: []coreutils.Technology{coreutils.Maven, coreutils.Npm}}, want: func() {
+			assert.Equal(t, "", os.Getenv(utils.JfPackageManagerEnvVariable))
+			assert.Equal(t, "", os.Getenv(utils.JfLanguageEnvVariable))
+			assert.Equal(t, "msi", os.Getenv(utils.JfMsiEnvVariable))
+		}},
```
Comment (attiasas) [inline commands/audit/jas/common_test.go:113]: add tests for Zero tech

File: commands/audit/audit.go [hunk 21]
```diff
@@ -79,6 +86,16 @@ func (auditCmd *AuditCommand) CreateXrayGraphScanParams() *services.XrayGraphSca
...
 	params.IncludeLicenses = auditCmd.IncludeLicenses
+	params.MultiScanId = auditCmd.analyticsMetricsService.GetMsi()
+	xscManager := auditCmd.analyticsMetricsService.XscManager()
+	if xscManager != nil {
+		version, err := xscManager.GetVersion()
+		if err != nil {
+			log.Debug(fmt.Sprintf("Can't get XSC version for xray graph scan params. Cause: %s", err.Error()))
+		}
+		params.XscVersion = version
+	}
```
Comment (attiasas) [inline commands/audit/audit.go:97]:
```suggestion
	if xscManager := auditCmd.analyticsMetricsService.XscManager(); xscManager != nil {
		version, err := xscManager.GetVersion()
		if err != nil {
			log.Debug(fmt.Sprintf("Can't get XSC version for xray graph scan params. Cause: %s", err.Error()))
		}
		params.XscVersion = version
	}
```

File: tests/utils/test_utils.go (hunk 16 above)
Comment (attiasas) [reply tests/utils/test_utils.go:25]: No need to prepare to that right now, when we will need it, we will change.
This addes more complexity where it is not needed right now

File: tests/utils/test_validation.go [hunk 22]
```diff
@@ -26,6 +26,18 @@ func ValidateXrayVersion(t *testing.T, minVersion string) {
...
+func ValidateXscVersion(t *testing.T, minVersion string) {
+	xscVersion, err := getXscVersion()
+	if err != nil {
+		assert.NoError(t, err)
+		return
+	}
+	err = clientUtils.ValidateMinimumVersion(clientUtils.Xsc, xscVersion.GetVersion(), minVersion)
+	if err != nil {
+		t.Skip(err)
+	}
```
Comment (attiasas) [inline tests/utils/test_validation.go:38]:
```suggestion
	if err = clientUtils.ValidateMinimumVersion(clientUtils.Xsc, xscVersion.GetVersion(), minVersion); err != nil {
		t.Skip(err)
	}
```

File: utils/analyticsmetrics.go [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+func (ams *AnalyticsMetricsService) CreateGeneralEvent(product xscservices.ProductName, eventType xscservices.EventType) *xscservices.XscAnalyticsGeneralEvent {
+	osAndArc, err := coreutils.GetOSAndArc()
+	curOs, curArch := "", ""
+	if err != nil {
+		log.Debug(fmt.Errorf("failed to get os and arcitucture for general event request to XSC service, error: %s ", err.Error()))
+	} else {
+		splitOsAndArch := strings.Split(osAndArc, "-")
+		curOs = splitOsAndArch[0]
+		curArch = splitOsAndArch[1]
+	}
```
Comment (attiasas) [inline utils/analyticsmetrics.go:96]:
```suggestion
	curOs, curArch := "", ""
	if osAndArc, err := coreutils.GetOSAndArc(); err != nil {
		log.Debug(fmt.Errorf("failed to get os and arcitucture for general event request to XSC service, error: %s ", err.Error()))
	} else {
		splitOsAndArch := strings.Split(osAndArc, "-")
		curOs = splitOsAndArch[0]
		curArch = splitOsAndArch[1]
	}
```

File: utils/analyticsmetrics.go [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+
+	event := xscservices.XscAnalyticsGeneralEvent{
+		XscAnalyticsBasicGeneralEvent: xscservices.XscAnalyticsBasicGeneralEvent{
+			EventType:              eventType,
+			EventStatus:            xscservices.Started,
+			Product:                product,
+			JfrogUser:              ams.xscManager.Config().GetServiceDetails().GetUser(),
+			OsPlatform:             curOs,
+			OsArchitecture:         curArch,
+			AnalyzerManagerVersion: GetAnalyzerManagerVersion(),
```
Comment (attiasas) [inline utils/analyticsmetrics.go:106]: GetAnalyzerManagerVersion will soon return `[RELEASE]` default value. I think it is neccessery to implement running the command `version` on the analyzerManager binary.

File: utils/analyticsmetrics.go [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+		},
+	}
+	return &event
```
Comment (attiasas) [inline utils/analyticsmetrics.go:109]:
```suggestion
	return &xscservices.XscAnalyticsGeneralEvent{
		XscAnalyticsBasicGeneralEvent: xscservices.XscAnalyticsBasicGeneralEvent{
			EventType:              eventType,
			EventStatus:            xscservices.Started,
			Product:                product,
			JfrogUser:              ams.xscManager.Config().GetServiceDetails().GetUser(),
			OsPlatform:             curOs,
			OsArchitecture:         curArch,
			AnalyzerManagerVersion: GetAnalyzerManagerVersion(),
		},
	}
```

File: utils/analyticsmetrics.go [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+func (ams *AnalyticsMetricsService) AddGeneralEvent(event *xscservices.XscAnalyticsGeneralEvent) {
+	if !ams.ShouldReportEvents() {
+		log.Debug("A general event request was not sent to XSC - analytics metrics are disabled.")
```
Comment (attiasas) [inline utils/analyticsmetrics.go:114]:
```suggestion
		log.Debug("Analytics metrics are disabled, skipping sending event request to XSC")
```

File: utils/analyticsmetrics.go [hunk 27]
```diff
@@ -0,0 +1,204 @@
...
+	// Set event's analytics data.
+	ams.SetMsi(msi)
+	ams.SetStartTime()
+}
+
+func (ams *AnalyticsMetricsService) UpdateGeneralEvent(event *xscservices.XscAnalyticsGeneralEventFinalize) {
+	if !ams.ShouldReportEvents() {
+		log.Debug("A general event update request was not sent to XSC - analytics metrics are disabled.")
+		return
+	}
```
Comment (attiasas) [inline utils/analyticsmetrics.go:132]:
```suggestion
	if !ams.ShouldReportEvents() {
		return
	}
```
No need to log at the end, we already logged that it is disabled

File: utils/analyticsmetrics.go [hunk 28]
```diff
@@ -0,0 +1,204 @@
...
+	if ams.msi == "" {
+		log.Debug("A general event update request was not sent to XSC - failed to add the starting event, and there is no valid multi-scan identifier provided.")
```
Comment (attiasas) [inline utils/analyticsmetrics.go:134]:
```suggestion
		log.Debug("MultiScanId is empty, skipping update general event...")
```

File: utils/analyticsmetrics.go [hunk 29]
```diff
@@ -0,0 +1,204 @@
...
+		return
+	}
+	err := ams.xscManager.UpdateAnalyticsGeneralEvent(*event)
+	if err != nil {
+		log.Debug(fmt.Sprintf("failed updading general event request in XSC service for multi_scan_id %s, error: %s \"", ams.GetMsi(), err.Error()))
+	}
```
Comment (attiasas) [inline utils/analyticsmetrics.go:140]:
```suggestion
	if err := ams.xscManager.UpdateAnalyticsGeneralEvent(*event); err != nil {
		log.Debug(fmt.Sprintf("failed updading general event request in XSC service for multi_scan_id %s, error: %s \"", ams.GetMsi(), err.Error()))
	} else {
        log.Debug(fmt.Sprintf("General event updated\n%s", *event)
    }
```
We can show the summary information on debug level.

File: utils/analyticsmetrics.go [hunk 30]
```diff
@@ -0,0 +1,204 @@
...
+func (ams *AnalyticsMetricsService) GetGeneralEvent(msi string) (*xscservices.XscAnalyticsGeneralEvent, error) {
+	if !ams.ShouldReportEvents() {
+		log.Debug("Can't get general event from XSC - analytics metrics are disabled.")
+		return nil, nil
+	}
+	event, err := ams.xscManager.GetAnalyticsGeneralEvent(msi)
+	if err != nil {
+		log.Debug(fmt.Sprintf("failed getting general event from XSC service for multi_scan_id %s, error: %s \"", msi, err.Error()))
+	}
+	return event, err
```
Comment (attiasas) [inline utils/analyticsmetrics.go:152]:
```suggestion
	return ams.xscManager.GetAnalyticsGeneralEvent(msi)
```
log can be added where this func is used, we return it...

File: utils/analyticsmetrics.go [hunk 31]
```diff
@@ -0,0 +1,204 @@
...
+		TotalScanDuration: totalDuration.String(),
+	}
+	return &xscservices.XscAnalyticsGeneralEventFinalize{
+		MultiScanId:                   ams.msi,
+		XscAnalyticsBasicGeneralEvent: basicEvent,
+	}
+}
+
+// Counts the total amount of findings in the provided results and updates the AnalyticsMetricsService with the amount of the new added findings
+func (ams *AnalyticsMetricsService) CountScanResultsFindings(scanResults *Results) int {
```
Comment (attiasas) [inline utils/analyticsmetrics.go:174]: I would consider adding this logic as methods in the `Results` struct

File: utils/mocks.go [hunk 32]
```diff
@@ -0,0 +1,51 @@
+package utils
```
Comment (attiasas) [inline utils/mocks.go:1]: move this file to `tests/utils` because it contains `t *testing.T` please rename this file to `test_mocks.go`

File: go.mod [hunk 33]
```diff
@@ -98,8 +98,10 @@ require (
 	gopkg.in/warnings.v0 v0.1.2 // indirect
 )
 
-// replace github.com/jfrog/jfrog-cli-core/v2 => github.com/jfrog/jfrog-cli-core/v2 dev
+replace github.com/jfrog/jfrog-cli-core/v2 => github.com/gailazar300/jfrog-cli-core/v2 v2.0.0-20240403103644-317f145c1f66
 
-// replace github.com/jfrog/jfrog-client-go => github.com/jfrog/jfrog-client-go dev
+//replace github.com/jfrog/jfrog-cli-core/v2 => ../jfrog-cli-core
```
Comment (attiasas) [inline go.mod:103]:
```suggestion
replace github.com/jfrog/jfrog-cli-core/v2 => github.com/jfrog/jfrog-cli-core/v2 dev

replace github.com/jfrog/jfrog-client-go => github.com/jfrog/jfrog-client-go dev
```
Update to dev branch, those changes already merged

File: xsc_test.go [hunk 34]
```diff
@@ -0,0 +1,47 @@
...
```
Comment (attiasas) [inline xsc_test.go:47]: No need for specific file for analytics integration and add another scan for that. it can be added to an existing test at `audit_test.go` for example `TestXrayAuditJasSimpleJson` since it will also have JAS issues. the func `validateAnalyticsBasicEvent` can be moved to `tests/utils/test_validation.go`

**Review by attiasas — COMMENTED** [review]
Please make sure it is working on platform that don't have XSC

File: tests/utils/test_validation.go (hunk 22 above)
Comment (gailazar300) [reply tests/utils/test_validation.go:38]: The way I implemented it is totally fine as well.

File: commands/audit/audit.go (hunk 21 above)
Comment (gailazar300) [reply commands/audit/audit.go:97]: The way I implemented it is totally fine as well.

File: utils/analyticsmetrics.go (hunk 23 above)
Comment (gailazar300) [reply utils/analyticsmetrics.go:96]: its the same.

File: utils/analyticsmetrics.go (hunk 25 above)
Comment (gailazar300) [reply utils/analyticsmetrics.go:109]: The way I implemented it is totally fine as well.

File: utils/analyticsmetrics.go (hunk 24 above)
Comment (gailazar300) [reply utils/analyticsmetrics.go:106]: The person who will implement it in the future will do so when it is relevant.

File: utils/analyticsmetrics.go (hunk 27 above)
Comment (gailazar300) [reply utils/analyticsmetrics.go:132]: I think that its ok to add in debug mode also if we are skipping the updating.

File: utils/analyticsmetrics.go (hunk 30 above)
Comment (gailazar300) [reply utils/analyticsmetrics.go:152]: this function is for testing and getting the analytics from xsc - I think that its fine like this.

File: utils/mocks.go (hunk 32 above)
Comment (gailazar300) [reply utils/mocks.go:1]: this caused circle import.
the unit tests are using this mock, so this struct can be in theire directory.

File: go.mod (hunk 33 above)
Comment (gailazar300) [reply go.mod:103]: again- of course, I wont merge the go.mod like this!
and will run tests with the dev branches, no need to write it every time.

File: xsc_test.go (hunk 34 above)
Comment (gailazar300) [reply xsc_test.go:47]: again as I mentioned before - I think that it's necessary to have a separate section for XSC testing and to have a test that tests analytics and nothing else.
I don't think that adding this in the audit section in one random test and validating analytics after other things will be right and clear.

Comment (gailazar300) [reply xsc_test.go:47]: and I will add here also about "InitSecurityTest" - we are going to work mostly with XSC instead of xray for now on - so now when the infrastructure is still small it's time to add testing xsc support.

**Review by attiasas — APPROVED** [review]
